File: crates/somite-ops/src/nfcore.rs
//! Official nf-core discovery behind one cache-friendly operator interface.

use std::collections::BTreeMap;
use std::process::Command;

use somite_ir::{ParamValue, PortType};

use crate::{Cost, OpKind, Operator, OutputSpec, ParamSpec, PortSpec, PortsSpec};

pub const CATALOG_URL: &str = "https://nf-co.re/pipelines.json";
pub type FetchResult = Result<(String, Vec<Pipeline>), String>;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Pipeline {
    pub name: String,
    pub description: String,
    pub topics: Vec<String>,
    pub revision: String,
}

impl Pipeline {
    pub fn operator_id(&self) -> String {
        format!("nf.{}", self.name)
    }

    pub fn operator(&self) -> Operator {
        let mut params = BTreeMap::new();
        params.insert(
            "revision".into(),
            ParamSpec {
                ty: "string".into(),
                label: Some("Version".into()),
                page: Some("Pipeline".into()),
                default: Some(ParamValue::String(self.revision.clone())),
                required: true,
                min: None,
                max: None,
            },
        );
        let ports = PortsSpec {
            r#in: vec![PortSpec {
                name: "sheet".into(),
                ty: PortType::Table,
                union: Vec::new(),
                optional: true,
                resource: None,
                stage_as: None,
                import_param: None,
            }],
            out: vec![PortSpec {
                name: "results".into(),
                ty: PortType::Directory,
                union: Vec::new(),
                optional: true,
                resource: None,
                stage_as: None,
                import_param: None,
            }],
        };
        let mut outputs = BTreeMap::new();
        outputs.insert(
            "results".into(),
            OutputSpec {
                glob: "{work}/out".into(),
                ty: PortType::Directory,
                optional: true,
                exclude: Vec::new(),
            },
        );
        Operator {
            id: self.operator_id(),
            title: format!("nf-core/{}", self.name),
            palette: vec!["nf-core".into(), "Catalog".into()],
            kind: OpKind::External,
            cost: Cost::High,
            bin: Some("nextflow".into()),
            pixi: vec!["bioconda::nextflow".into()],
            params,
            ports,
            argv: vec![
                "nextflow".into(),
                "run".into(),
                format!("nf-core/{}", self.name),
                "-r".into(),
                "{param.revision}".into(),
                "--input".into(),
                "{input.sheet}".into(),
                "--outdir".into(),
                "{work}/out".into(),
            ],
            outputs,
            stdout: None,
            resolution: None,
        }
    }
}

pub fn parse(text: &str) -> Result<Vec<Pipeline>, String> {
    let root: serde_json::Value = serde_json::from_str(text).map_err(|error| error.to_string())?;
    let workflows = root
        .get("remote_workflows")
        .and_then(serde_json::Value::as_array)
        .ok_or_else(|| "nf-core catalog has no remote_workflows array".to_owned())?;
    let mut pipelines = workflows
        .iter()
        .filter(|workflow| {
            !workflow
                .get("archived")
                .and_then(serde_json::Value::as_bool)
                .unwrap_or(false)
        })
        .filter_map(|workflow| {
            let name = workflow.get("name")?.as_str()?.to_owned();
            let revision = workflow
                .get("releases")?
                .as_array()?
                .iter()
                .filter_map(|release| release.get("tag_name")?.as_str())
                .find(|tag| *tag != "dev")?
                .to_owned();
            let description = workflow
                .get("description")
                .and_then(serde_json::Value::as_str)
                .unwrap_or_default()
                .to_owned();
            let topics = workflow
                .get("topics")
                .and_then(serde_json::Value::as_array)
                .into_iter()
                .flatten()
                .filter_map(serde_json::Value::as_str)
                .map(str::to_owned)
                .collect();
            Some(Pipeline {
                name,
                description,
                topics,
                revision,
            })
        })
        .collect::<Vec<_>>();
    pipelines.sort_by(|left, right| left.name.cmp(&right.name));
    Ok(pipelines)
}

pub fn fetch() -> FetchResult {
    let output = Command::new("curl")
        .args(["-fsSL", "--max-time", "15", CATALOG_URL])
        .output()
        .map_err(|error| format!("could not start curl: {error}"))?;
    if !output.status.success() {
        return Err(format!("nf-core catalog returned {}", output.status));
    }
    let text = String::from_utf8(output.stdout).map_err(|error| error.to_string())?;
    let pipelines = parse(&text)?;
    Ok((text, pipelines))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_current_shape_and_ignores_archived_or_unreleased_entries() {
        let text = r#"{
          "remote_workflows": [
            {"name":"rnaseq","description":"RNA analysis","topics":["rna"],"archived":false,
             "releases":[{"tag_name":"3.26.0"}]},
            {"name":"old","archived":true,"releases":[{"tag_name":"1.0.0"}]},
            {"name":"dev-only","archived":false,"releases":[{"tag_name":"dev"}]},
            {"name":"new","archived":false,"releases":[]}
          ]
        }"#;

        let pipelines = parse(text).expect("valid catalog");
        assert_eq!(pipelines.len(), 1);
        assert_eq!(pipelines[0].name, "rnaseq");
        assert_eq!(pipelines[0].revision, "3.26.0");
        assert_eq!(pipelines[0].operator().argv[2], "nf-core/rnaseq");
    }
}

File: packages/workflow/nfcore.ts
import type { Operator, PinnedOperator } from "./catalog.ts";
import { operatorRevision } from "./catalogCodec.ts";

export const NFCORE_CATALOG_URL = "https://nf-co.re/pipelines.json";

export type NfcorePipeline = Readonly<{
  name: string;
  description: string;
  topics: readonly string[];
  revision: string;
  resolvedRevision: string;
}>;

export type NfcoreCatalogResponse = Readonly<{
  entries: readonly Readonly<{
    operator: PinnedOperator;
    description: string;
    topics: readonly string[];
    revision: string;
  }>[];
  cached: boolean;
}>;

function object(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : undefined;
}

function string(value: unknown) {
  return typeof value === "string" && value ? value : undefined;
}

function strings(value: unknown) {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
}

export function parseNfcoreCatalog(value: string | unknown): NfcorePipeline[] {
  const root = object(typeof value === "string" ? JSON.parse(value) : value);
  if (!Array.isArray(root?.remote_workflows)) throw new Error("nf-core catalog has no remote_workflows array");
  const pipelines = root.remote_workflows.flatMap((candidate): NfcorePipeline[] => {
    const workflow = object(candidate);
    if (!workflow || workflow.archived === true) return [];
    const name = string(workflow.name);
    if (!name || !/^[A-Za-z0-9_-]+$/.test(name)) return [];
    const releases = Array.isArray(workflow.releases) ? workflow.releases.map(object).filter((release): release is Record<string, unknown> => Boolean(release)) : [];
    const release = releases.find((item) => string(item.tag_name) !== "dev" && string(item.tag_name) && /^[A-Za-z0-9._-]+$/.test(string(item.tag_name)!));
    const revision = string(release?.tag_name);
    const resolvedRevision = string(release?.tag_sha);
    if (!revision || !resolvedRevision || !/^[0-9a-f]{40}$/.test(resolvedRevision)) return [];
    return [{
      name,
      description: string(workflow.description) ?? "",
      topics: strings(workflow.topics),
      revision,
      resolvedRevision,
    }];
  });
  pipelines.sort((left, right) => left.name.localeCompare(right.name));
  return pipelines;
}

export function nfcoreOperator(pipeline: NfcorePipeline): PinnedOperator {
  const operator: Operator = {
    id: `nf.${pipeline.name}`,
    title: `nf-core/${pipeline.name}`,
    palette: ["nf-core", "Catalog"],
    kind: "external",
    cost: "high",
    bin: "nextflow",
    pixi: ["bioconda::nextflow"],
    params: {
      revision: { type: "string", label: "Version", page: "Pipeline", default: pipeline.revision, required: true },
    },
    ports: {
      in: [{ name: "sheet", type: "Table", optional: true }],
      out: [{ name: "results", type: "Directory", optional: true }],
    },
    argv: ["nextflow", "run", `nf-core/${pipeline.name}`, "-r", "{param.revision}", "--input", "{input.sheet}", "--outdir", "{work}/out"],
    outputs: { results: { glob: "{work}/out", type: "Directory", optional: true } },
  };
  return { ...operator, revision: operatorRevision(operator) };
}

export function nfcoreCatalogResponse(pipelines: readonly NfcorePipeline[], cached: boolean): NfcoreCatalogResponse {
  return {
    entries: pipelines.map((pipeline) => ({
      operator: nfcoreOperator(pipeline),
      description: pipeline.description,
      topics: pipeline.topics,
      revision: pipeline.revision,
    })),
    cached,
  };
}

export function searchNfcoreCatalog(pipelines: readonly NfcorePipeline[], query: string, limit = 12) {
  const normalized = query.trim().replace(/\s+/g, " ");
  if (!normalized || normalized.length > 120 || [...normalized].some((character) => {
    const code = character.codePointAt(0) ?? 0;
    return code < 32 || code === 127;
  })) {
    throw new Error("nf-core query must contain 1 to 120 printable characters");
  }
  const terms = normalized.toLocaleLowerCase("en-US").split(" ");
  const matches = pipelines.filter((pipeline) => {
    const haystack = `${pipeline.name} ${pipeline.description} ${pipeline.topics.join(" ")}`.toLocaleLowerCase("en-US");
    return terms.every((term) => haystack.includes(term));
  });
  return {
    query: normalized,
    provenance: NFCORE_CATALOG_URL,
    total_matches: matches.length,
    entries: matches.slice(0, Math.max(1, Math.min(50, limit))).map((pipeline) => ({
      repository: `nf-core/${pipeline.name}`,
      title: `nf-core/${pipeline.name}`,
      description: pipeline.description,
      topics: pipeline.topics,
      revision: pipeline.revision,
    })),
  };
}
